File: Interpreter/src/scanner/token/token.go
package token

import (
	"fmt"
)

type Token struct {
	line   int
	column int
	text   string
	class  Token_Class
}

type Token_Class int

const (
	IDENTIFIER         Token_Class = iota // 'my_var'
	PERIOD                                // '.'
	QUESTION                              // '?'
	COMMA                                 // ','
	COLON                                 // ':'
	SEMICOLON                             // ';'
	BIT_AND                               // '&'
	BIT_OR                                // '|'
	ADD_OP                                // '+'
	MIN_OP                                // '-'
	MULT_OP                               // '*'
	DIV_OP                                // '/'
	MOD_OP                                // '%'
	INCREMENT                             // '++'
	DECREMENT                             // '--'
	POWER_OP                              // '**'
	LESS_THAN                             // '<'
	GREATER_THAN                          // '>'
	ASSIGN                                // '='
	LESS_THAN_EQ                          // '<='
	GREATER_THAN_EQ                       // '>='
	EQUAL                                 // '=='
	PLUS_EQ                               // '+='
	MIN_EQ                                // '-='
	TIMES_EQ                              // '*='
	DIV_EQ                                // '/='
	MOD_EQ                                // '%='
	OR                                    // '||'
	AND                                   // '&&'
	RANGE                                 // '..'
	OR_EQUAL                              // '|='
	LSHOVEL                               // '<<'
	RSHOVEL                               // '>>'
	LCURLY                                // '{'
	RCURLY                                // '}'
	LPAREN                                // '('
	RPAREN                                // ')'
	LBRACKET                              // '['
	RBRACKET                              // ']'
	STRING_LITERAL                        // "mystring" or 'mystring'
	NUM_LITERAL                           // '0435.345'
	MULTILINE_COMMENT                     // '/* some comment */'
	SINGLELINE_COMMENT                    // '// some comment '
	FUNC_KEYWORD                          // 'function'
	SPAWN_KEYWORD                         // 'spawn'
	KEYWORD                               // 'do end for function...etc'
)

var KeywordMap = map[string]Token_Class{
	"function": FUNC_KEYWORD,
	"spawn":    SPAWN_KEYWORD,
}

func New(val string, class Token_Class, line int, col int) *Token {
	return &Token{line, col - len(val), val, class}
}

func (this *Token) String() string {
	return fmt.Sprintf("{ line: %d, column: %d, val: %s, class: %s }", this.line, this.column, this.text, this.classString())
}

func (this *Token) classString() string {
	switch this.class {
	case IDENTIFIER:
		return "IDENTIFIER"
	case PERIOD:
		return "PERIOD"
	case QUESTION:
		return "QUESTION"
	case COMMA:
		return "COMMA"
	case COLON:
		return "COLON"
	case SEMICOLON:
		return "SEMICOLON"
	case BIT_AND:
		return "BIT_AND"
	case BIT_OR:
		return "BIT_OR"
	case ADD_OP:
		return "ADD_OP"
	case MIN_OP:
		return "MIN_OP"
	case MULT_OP:
		return "MULT_OP"
	case DIV_OP:
		return "DIV_OP"
	case MOD_OP:
		return "MOD_OP"
	case INCREMENT:
		return "INCREMENT"
	case DECREMENT:
		return "DECREMENT"
	case POWER_OP:
		return "POWER_OP"
	case LESS_THAN:
		return "LESS_THAN"
	case GREATER_THAN:
		return "GREATER_THAN"
	case ASSIGN:
		return "ASSIGN"
	case LESS_THAN_EQ:
		return "LESS_THAN_EQ"
	case GREATER_THAN_EQ:
		return "GREATER_THAN_EQ"
	case EQUAL:
		return "EQUAL"
	case PLUS_EQ:
		return "PLUS_EQ"
	case MIN_EQ:
		return "MIN_EQ"
	case TIMES_EQ:
		return "TIMES_EQ"
	case DIV_EQ:
		return "DIV_EQ"
	case MOD_EQ:
		return "MOD_EQ"
	case OR:
		return "OR"
	case AND:
		return "AND"
	case OR_EQUAL:
		return "OR_EQUAL"
	case LSHOVEL:
		return "LSHOVEL"
	case RSHOVEL:
		return "RSHOVEL"
	case LCURLY:
		return "LCURLY"
	case RCURLY:
		return "RCURLY"
	case LPAREN:
		return "LPAREN"
	case RPAREN:
		return "RPAREN"
	case LBRACKET:
		return "LBRACKET"
	case RBRACKET:
		return "RBRACKET"
	case STRING_LITERAL:
		return "STRING_LITERAL"
	case NUM_LITERAL:
		return "NUM_LITERAL"
	case KEYWORD:
		return "KEYWORD"
	case MULTILINE_COMMENT:
		return "MULTILINE_COMMENT"
	case SINGLELINE_COMMENT:
		return "SINGLELINE_COMMENT"
	case RANGE:
		return "RANGE"
	}
	return ""
}

File: rust/src/lib/lib_hy_syntax/token.rs
use std::fmt::{Show, Formatter, Result};
use std::char;

#[deriving(Clone, Eq, Show)]
pub enum TokenType {
    Identifier,
    Period,
    Bang,
    Question,
    Comma,
    Colon,
    Semicolon,
    Bit_And,
    Bit_Or,
    Bit_Negate,
    Add_Op,
    Min_Op,
    Mult_Op,    
    Div_Op,
    Mod_Op,
    Power_Op,
    Increment,
    Decrement,
    Less_Than,
    Greater_Than,
    Assign,
    Less_Than_Eq,
    Greater_Than_Eq,
    Equal,
    Bang_Eq,
    Plus_Eq,
    Min_Eq,
    Times_Eq,
    Div_Eq,
    Mod_Eq,
    Power_Eq,
    Bit_Or_Eq,
    Logical_Or_Eq,
    Bit_And_Eq,
    Logical_Or,
    Logical_And,
    Excl_Range,
    Incl_Range,
    Lshovel,
    Rshovel,
    Lcurly,
    Rcurly,
    Lparen,
    Rparen,
    Lbracket,
    Rbracket,
    Chan_Recv,
    Chan_Send,
    Pub_Class_Var,
    Priv_Class_Var,
    Pub_Inst_Var,
    Priv_Inst_Var,
    Function,
    Spawn,
    If,
    Else,
    Then,
    For,
    In,
    While,
    Do,
    Not,
    And,
    Or,
    From,
    To,
    By,
    End,
    Class,
    Import,
    As,
    Export,
    Super,
    Supers,
    This,
    Return,
    Operator,
    Extends,
    Generator,
    Either,
    Wait_For,
    Break,
    Continue,
    Yield,
    Given,
    Is,
    Var,
    True,
    False,
    New,
    Upto,
    Through,
    String_Literal,
    Binary_Literal,
    Hex_Literal,
    Int_Literal,
    Float_Literal,
    Multiline_Comment,
    Singleline_Comment,
    NewLine
}

#[deriving(Clone, Eq)]
pub struct Token {
    pub text       : ~str,
    pub typ        : TokenType,
    pub line       : uint,
    pub col        : uint,
    pub buf_offset : uint
}

impl Token {
    pub fn start(self) -> uint {
        self.buf_offset
    }

    pub fn end(self) -> uint {
        self.buf_offset + self.text.as_bytes().len()
    }
}

impl Show for Token {
    fn fmt(&self, f: &mut Formatter) -> Result {
        let info = format!("{:?} at {}:{}", self.typ, self.line, self.col);
        write!(f, "\\{\n  {}\n  text: '{}'\n\\}", info, self.text)
    }
}

pub fn str_to_type(tok_str : &str) -> TokenType {
    match tok_str {
        "." => Period,
        "!" => Bang,
        "?" => Question,
        "," => Comma,
        ":" => Colon,
        ";" => Semicolon,
        "&" => Bit_And,
        "|" => Bit_Or,
        "~" => Bit_Negate,
        "+" => Add_Op,
        "-" => Min_Op,
        "*" => Mult_Op,
        "/" => Div_Op,
        "%" => Mod_Op,
        "^" => Power_Op,
        "++" => Increment,
        "--" => Decrement,
        "<" => Less_Than,
        ">" => Greater_Than,
        "=" => Assign,
        "<=" => Less_Than_Eq,
        ">=" => Greater_Than_Eq,
        "==" => Equal,
        "!=" => Bang_Eq,
        "+=" => Plus_Eq,
        "-=" => Min_Eq,
        "*=" => Times_Eq,
        "/=" => Div_Eq,
        "%=" => Mod_Eq,
        "^=" => Power_Eq,
        "|=" => Bit_Or_Eq,
        "||=" => Logical_Or_Eq,
        "&=" => Bit_And_Eq,
        "||" => Logical_Or,
        "&&" => Logical_And,
        ".." => Excl_Range,
        "..." => Incl_Range,
        "<<" => Lshovel,
        ">>" => Rshovel,
        "{" => Lcurly,
        "}" => Rcurly,
        "(" => Lparen,
        ")" => Rparen,
        "[" => Lbracket,
        "]" => Rbracket,
        "<-" => Chan_Recv,
        "->" => Chan_Send,
        "#" => Pub_Class_Var,
        "#_" => Priv_Class_Var,
        "@" => Pub_Inst_Var,
        "@_" => Priv_Inst_Var,                 
        "function" => Function,
        "spawn" => Spawn,
        "if" => If,
        "else" => Else,
        "then" => Then,
        "for" => For,
        "in" => In,
        "while" => While,
        "do" => Do,
        "not" => Not,
        "and" => And,
        "or" => Or,
        "from" => From,
        "to" => To,
        "by" => By,
        "end" => End,
        "class" => Class,
        "import" => Import,
        "as" => As,
        "export" => Export,
        "super" => Super,
        "supers" => Supers,
        "this" => This,
        "return" => Return,
        "operator" => Operator,
        "extends" => Extends,
        "gen" => Generator,
        "either" => Either,
        "wait_for" => Wait_For,
        "break" => Break,
        "continue" => Continue,
        "yield" => Yield,
        "given" => Given,
        "is" => Is,
        "var" => Var,
        "true" => True,
        "false" => False,
        "new" => New,
        "upto" => Upto,
        "through" => Through,
        "\n" => NewLine,
        _ if tok_str.starts_with("'") => String_Literal,
        _ if tok_str.starts_with("\"") => String_Literal,
        _ if tok_str.starts_with("0b") => Binary_Literal,
        _ if tok_str.starts_with("0B") => Binary_Literal,
        _ if tok_str.starts_with("0x") => Hex_Literal,
        _ if tok_str.starts_with("0X") => Hex_Literal,
        _ if tok_str.starts_with("/*") => Multiline_Comment,
        _ if tok_str.starts_with("//") => Singleline_Comment,  
        _ if char::is_digit(tok_str.char_at(0)) => {
            if tok_str.contains_char('.') {
                Float_Literal
            } else {
                Int_Literal
            }
        },                                
        _ => Identifier               
    }     
}